egs/wsj/s5/steps/conf/train_calibration.sh

```diff
@@ -0,0 +1,122 @@
+  echo " Options:"
+  echo "    --cmd (run.pl|queue.pl...)      # specify how to run the sub-processes."
+  echo "    --lmwt <int>                    # scaling for confidence extraction"
+  echo "    --decode-mbr <bool>             # use Minimum Bayes Risk decoding"
+  echo "    --grep-filter <str>             # remove words from calibration targets"
+  exit 1;
+fi
+
+set -euxo pipefail
+
+data=$1
+lang=$2 # Note: may be graph directory not lang directory, but has the necessary stuff copied.
+unigrams=$3
+latdir=$4
+dir=$5
+
+model=$latdir/../final.mdl # assume model one level up from decoding dir.
+
+for f in $data/text $lang/words.txt $unigrams $latdir/lat.1.gz; do
+  [ ! -f $f ] && echo "$0: Missing file $f" && exit 1
+done
+[ -z "$cmd" ] && echo "$0: Missing --cmd '...'" && exit 1
+
+[ -d $dir/log ] || mkdir -p $dir/log
+nj=$(cat $latdir/num_jobs)
+
+# Store the setup,
+echo $lmwt >$dir/lmwt
+echo $decode_mbr >$dir/decode_mbr
+cp $unigrams $dir/unigrams
+
+# Create the ctm with raw confidences,
+# - we keep the timing relative to the utterance,
+if [ $stage -le 0 ]; then
+  $cmd JOB=1:$nj $dir/log/get_ctm.JOB.log \
+    lattice-scale --inv-acoustic-scale=$lmwt "ark:gunzip -c $latdir/lat.JOB.gz|" ark:- \| \
+    lattice-limit-depth ark:- ark:- \| \
+    lattice-push --push-strings=false ark:- ark:- \| \
+    lattice-align-words-lexicon --max-expand=10.0 \
+     $lang/phones/align_lexicon.int $model ark:- ark:- \| \
+    lattice-to-ctm-conf --decode-mbr=$decode_mbr ark:- - \| \
+    utils/int2sym.pl -f 5 $lang/words.txt \
+    '>' $dir/JOB.ctm
+  # Merge and clean,
+  set +x; for ((n=1; n<=nj; n++)); do cat $dir/${n}.ctm; done > $dir/ctm
+  rm $dir/*.ctm; set -x
+fi
+
+# Get evaluation of the 'ctm' using the 'text' reference,
+if [ $stage -le 1 ]; then
+  steps/conf/convert_ctm_to_tra.py $dir/ctm - | \
+  align-text --special-symbol="<eps>" ark:$data/text ark:- ark,t:- | \
+  utils/scoring/wer_per_utt_details.pl --special-symbol "<eps>" \
+  >$dir/align_text 
+  # Append alignment to ctm,
+  steps/conf/append_eval_to_ctm.py $dir/align_text $dir/ctm $dir/ctm_aligned
+fi
+
+# Prepare word-categories,
+steps/conf/prepare_word_categories.py --min-count $word_min_count $lang/words.txt $dir/ctm $dir/word_categories
+
+# Compute lattice-depth,
+latdepth=$dir/lattice_frame_depth.ark
+if [ $stage -le 2 ]; then
+  [ -e $latdepth ] || utils/lattice_depth_per_frame.sh --cmd "$cmd" $latdir $dir
+fi
+
+# Create the training data for logistic regression,
+if [ $stage -le 3 ]; then
+  steps/conf/prepare_calibration_data.py \
+    --conf-targets $dir/train_targets.ark --conf-feats $dir/train_feats.ark \
+    $dir/ctm_aligned $latdepth $unigrams $dir/word_categories
+fi
+
+# Filter the targets,
+if [ $stage -le 4 ]; then
+  # Removing words with format: [...], <...>, %..., ...-,
+  grep -i -v -E "\^(\[.*\]|<.*>|%[^,]*|[^,]*-)," $dir/train_targets.ark >$dir/train_targets_filt.ark
+fi
+
```

Comment:
Kare, it's problematic to do filtering of this type in anything but a script in local/.  This makes sense for English perhaps, given normal transcription conventions, but what about, say, Buckwalterized Arabic?  Some of these characters are a valid letter in that encoding.  All scripts are supposed to be language-independent and in general we don't do anything that you couldn't do in principle from the integerized form of the file, in any of the generic scripts.   In fact, using anything other than the integer identifiers in generic scripts  (except for scoring) is somewhere between highly discouraged to forbidden, because with encoding issues and locales you can run into a lot of unexpected issues.  I wonder if it would be possible to move the parts of this that need to look at words.txt to an outer level, so you'd have the user supply a file of the form

word-identifier1  feature1 feature2 .. featureN  
word-identifier2  feature1 feature2 .. featureN

and this way the script itself would never have to either read the actual words, or make language-specific assumptions?  This way you could add a feature for partial-words and the like, in a way that makes sense for the corpus, and add a feature for the length (and you could decide yourself whether you want to count the bytes or the Unicode characters), and so on.  This would also mean less work to do in prepare_calibration_data.py.